app/domain/authentication/Readme.md

```diff
@@ -0,0 +1,227 @@
+rescue ActiveSupport::Duration::ISO8601Parser::ParsingError
+  raise Errors::Authentication::DataObjects::InvalidTokenTTL.new(resource_id, @token_ttl)
+end
+```
+
+#### Authenticator Data Contract
+
+The data contract provides validation for data prior to initializing an Authenticator Data Object.
+
+Example: `Authentication::AuthnJwt::V2::DataObjects::AuthenticatorContract`
+
+Contracts are extended from the [Dry RB Validation library](https://dry-rb.org/gems/dry-validation/1.8/). They work by defining a schema:
+
+**Schemas**
+
+```rb
+module Authentication
+  module AuthnJwt
+    module V2
+      module DataObjects
+        class AuthenticatorContract < Dry::Validation::Contract
+
+          schema do
+            required(:account).value(:string)
+            required(:service_id).value(:string)
+
+            optional(:jwks_uri).value(:string)
+            optional(:public_keys).value(:string)
+            ...
+          end
+          ...
+        end
+      end
+    end
+  end
+end
+```
+
+which defines the required and optional data as well as the type. As Conjur Variables store values as strings, they type will always be `String`.
+
+**Validation Rules**
+
+With a schema defined, we can check data validity with rules:
+
+```ruby
+# Ensure claims has only one `:` in it
+rule(:claim_aliases) do
+  if (bad_claim = values[:claim_aliases].to_s.split(',').find { |item| item.count(':') != 1 })
+    key.failure(
+      **response_from_exception(
+        Errors::Authentication::AuthnJwt::ClaimAliasNameInvalidCharacter.new(bad_claim)
+      )
+    )
+  end
+end
+```
+
+These rules are executed, top to bottom, additively.
+
+Contracts return a Success or Failure response, with either the successful result or a list of errors. We are using some trickery to mimic the existing Exception driven workflow for validation. By calling `failure` with desired exception formatted with `response_from_exception`, we are defining the desired exception that should be raised. The `AuthenticatorRepository` will raise the first exception resulting from the contract validation.
+
+#### Strategy
+
+A strategy handles the external validation of the provided identity. It follows the Command class pattern.
+
+Example: `Authentication::AuthnJwt::V2::Strategy`
+
+At minimum, a Strategy requires the following methods
+
+```ruby
+# Initializer
+#
+# @param [Authenticator] authenticator - Authenticator Data Object that holds all relevant authenticator specific information.
+#
+# Note: additional dependencies should be defined as default parameters
+def initialize(authenticator:)
+  @authenticator = authenticator
+  ...
+end
+```
+
+```ruby
+# Verifies the validity of the contents of the provided request body and/or request parameters
+#
+# @param [String] request_body - authentication request body
+# @param [Hash] parameters - authentication request parameters
+#
+# @return something suitable for identifying a Conjur Role (usually a String or Hash)
```

Comment:
In the case of the JWT Authenticator, we return the relevant part of the JWT token. Initially I was thinking is is because we perform some extraction in the Identity Resolution phase. In the OIDC Authenticator, we extract the desired lookup attribute from the claim (and return that). 

Until you asked the question, I didn't think about the fact that the ["pre-filter" logic](https://github.com/cyberark/conjur/blob/049bc4c13b3bc30fef8355312caa268b765036dc/app/domain/authentication/authn_jwt/v2/resolve_identity.rb#L14-L45) probably should live in the Strategy (and Strategies need to return a single identity to be mapped to a Conjur Role). This allows us to keep the implementation details (JWT/OIDC/AWS IAM, etc.) in the Strategy.

This change also enables the above refactor (streaming role match) to be significantly simpler.

Reply:
Scratch part of this above statement.  There is key information in the JWT that needs to be available in the Identity Resolution phase (ex annotation matching).